File: Python/Easy/Minimum-Absolute-Difference-in-BST.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution(object):
    sortedList = []
    def getMinimumDifference(self, root):
        """
        :type root: TreeNode
        :rtype: int
        """
        self.sortedList = []
        self.inorder(root)
        min = self.sortedList[1]-self.sortedList[0]
        for i in range(2,len(self.sortedList)):
            if (self.sortedList[i]-self.sortedList[i-1]<min):
                min = self.sortedList[i]-self.sortedList[i-1]
        return min
        
    def inorder(self, node):
        if (node is None):
            return
        self.inorder(node.left)
        self.sortedList.append(node.val)
        self.inorder(node.right)

File: Java/Easy/Minimum-Absolute-Difference-in-BST.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    TreeNode previousNode;
    int min = Integer.MAX_VALUE;

    public int getMinimumDifference(TreeNode root) {
        previousNode = null;
        inorderTraversal(root);
        return min;
    }
    
    // get the nodes in sorted order, then get the min difference
    void inorderTraversal(TreeNode root) {
        if (root==null) {
            return;
        }        
        inorderTraversal(root.left);
        if (previousNode!=null) {
            min = Math.min(min, Math.abs(previousNode.val- root.val));
        }
        previousNode = root;
        inorderTraversal(root.right);
    }
}
